app/models/invite.rb

```diff
@@ -40,18 +51,30 @@ def user_doesnt_already_exist
     return if email.blank?
     user = Invite.find_user_by_email(email)
 
-    if user && user.id != self.user_id
+    if user && user.id != self.invited_users&.first&.user_id
       @email_already_exists = true
       errors.add(:email)
     end
   end
 
+  def is_invite_link?
+    max_redemptions_allowed > 1
+  end
+
   def redeemed?
-    redeemed_at.present?
+    if is_invite_link?
+      redemption_count >= max_redemptions_allowed
```

Comment:
I think we need to be very careful here with regards to concurrency. The `redemption_count` is accurate at the point where the invite was loaded. In that time, other Rails processes can also be updating the count which means we might burst the redemption limit when alot of people are redeeming invites at the same time.

Reply:
I see. What should we do here to add more safeguards?

Reply:
I'll probably go with a mutex around the redemption.

Reply:
Done.